### ISSUE TITLE
Customize the drush cache directory - avoid rebuild collisions

### ISSUE THREAD

**Issue by danepowell** [drush-ops/drush#3592]
Using Drush 9, if two processes try to clear the Drush cache at the same time, it creates a race condition that sometimes results in an error:
> Failed to remove directory "/home/foo/.drush/cache/factory": rmdir(/home/foo/.drush/cache/factory): Directory not empty

This is much more likely to happen on shared file systems such as Gluster or NFS, presumably because the cache clear takes 1-2 seconds and there's more time to produce a conflict.

This is easy to reproduce on a shared filesystem. Simply open two or three terminals and run `drush cc drush` in each one simultaneously.

I think it's a reasonable request for Drush cache clears to be multi-process safe, because these commands often happen automatically as part of a CI or deployment process, and these processes often _must_ run in parallel for efficiency.

If there's no way to make the cache clears more tolerant of slow file systems or running in parallel, it would at least help to have a configuration option to disable caching altogether, either permanently or on a request-by-request basis.

Note that other folks have reported the same error with Symfony cache clears, but it seems unlikely that Symfony is going to address this at the Filesystem level: https://github.com/symfony/symfony/issues/27578

**Comment by weitzman** [comment]
In Drush8 and below, you can set CACHE_PREFIX env variable on a request by request basis thus isolating that request from others.

**Comment by weitzman** [comment]
You would set the tmp environment variable for Drush9. See https://github.com/drush-ops/drush/blob/3353529a89a95107a957e5d04c57f982541b9a98/src/Config/DrushConfig.php#L39-L56. I haven't tried this so there may be difficulties.

**Comment by danepowell** [comment]
You can modify the TMP directory in order to change where the cache is stored (although only if you also remove write permissions on `~/.drush`, which introduces problems of its own), but there's no way to completely disable the cache as Drush will throw an exception.

Do you see any problems with adding a locking mechanism to the cache clear script at the point where it touches the filesystem?

**Comment by weitzman** [comment]
i'd have to see the code to judge if the complexity tradeoff is worthwhile. are we just talking about drush cache here or also guards around drupal cache also?

**Comment by greg-1-anderson** [comment]
Is there any specific reason why you have your cache on a shared filesystem? Generally caches should not be shared.

**Comment by rsaddington** [comment]
The issue we're having is when deploying multiple sites in a multi-site setup - each site includes a 'drush cc drush' as part of the deployment process to ensure any new drush commands/aliases are made available.  All the sites shared the same drush cache directory and depending on where they are in the deployment process will attempt to read/write to this common location hence causing race conditions.

Would an alternative approach be to create a drush cache directory for each site/uri running on the platform? 

e.g. a third candidate directory along the lines of:
/tmp/drush-<drush-user>/<site-name-or-uri>/cache

Would this be a viable solution?

**Comment by weitzman** [comment]
Yes, thats viable, and you can do that yourself with the CACHE_PREFIX in Drush8. Its awkward to do in Drush9 per https://github.com/drush-ops/drush/issues/3592#issuecomment-400427642. I'd love some suggestions for how this should work there.

**Comment by greg-1-anderson** [comment]
Couldn't we just implement something like `CACHE_PREFIX` in Drush 9?  Maybe check if an environment variable `DRUSH_CACHE_DIR` exists, and prefer that location if so.

I am uncertain whether we should segregate the cache based on the target site. On the surface, this does not seem like a good idea, as it involves a lot of unnecessary duplication that most Drush users will not need. I don't think that concurrent cache clearing is a common need. Perhaps those that need to do this could add locking to their scripts that call Drush. I agree that aspirationally speaking, Drush should be safe to call concurrently. I'd have to say that this request is fairly low on the priority list IMO, but perhaps PRs to improve this would be welcome.

Simply supporting `DRUSH_CACHE_DIR` would be pretty easy, though; I'd be happy to submit a candidate PR for that, if there's general agreement on whether that's the right approach & environment variable name.

**Comment by danepowell** [comment]
See the PR I just opened as a proof of concept for a stupidly simple semaphore system that would make cache clears concurrency-safe. I could also see an argument for putting this into the Symfony filesystem--I might suggest it there but I don't expect it will be well received.

Supporting a CACHE_DIR environment variable would be better than nothing, but it would be a pain to implement. For libraries like BLT that call Drush dozens or hundreds of times, every single call would have to be manually prefixed with this variables. And customers not using BLT would have to "just know" to prefix their calls with this variable. I don't see how it could be solved at the platform level (setting an environment variable wouldn't work because every call would still be using the same cache directory).

If we're not willing to make cache clears concurrency-safe at their root, then my preference would be for a DISABLE_CACHING environment variable that just disables caching completely and sidesteps the whole issue.

**Comment by weitzman** [comment]
I'm not so inclined to proceed with the semaphore. I think the only ones who experience this error are hosts who put Drush cache onto a shared filesystem. The only such host I know of is Acquia Cloud. I recommend that Acquia globally configure the cache dir to a local directory, for all users, by default. I'll write the DRUSH_CACHE_DIR PR to make that easy in drush9.

The way to disable caching would be DRUSH_CACHE_DIR=/dev/null.

**Comment by danepowell** [comment]
Okay thanks, I'll be happy to help test that PR when it's ready.

**Comment by weitzman** [comment]
PR posted to #3597

**Comment by danepowell** [comment]
Sadly this is not totally solved with #3597. @weitzman do you want to reopen this or create a new issue?

I've discovered that the error can still occur on non-shared file systems, it's just less frequent. I think I also have some more insight on the root cause that might guide a solution.

The root of the problem _appears_ to be that Drush writes to the cache directory in the process of bootstrapping itself even if it's being called specifically to clear the cache. This is what creates the race condition. Specifically, I saw Drush write globally-defined command files to its cache as the first step in a cache clear.

So if you have two drush cache clears running concurrently, one process might be deleting the cache while the other is still bootstrapping and writing to it. Because Symfony Filesystem's `remove()` function enumerates directory contents and then iteratively deletes them, this causes the error.

It seems to be that one or both of these solutions would be ideal:
1. Symfony Filesystem's `remove()` should handle the situation more gracefully where a file gets written to a directory that it's trying to delete. Although I'm not exactly sure what that solution would look like :) If you successfully delete a directory that's being written to by another process, it seems like it would just shift the error to the writing process. I wonder how native utilities like `rm -rf` handle this?
2. Drush should not write to the cache if it's being called specifically to clear its own cache `drush cc drush`. This seems like the easiest fix, and just kind of makes sense to me.

**Comment by weitzman** [comment]
2) would be pretty awkward to implement as we have not parsed the command when we are writing those cache files. The cache writes are due to command discovery which must happen early, for obvious reasons.

Can you try a strategy where you isolate each site in the multisite to its own cache?

**Comment by danepowell** [comment]
Yeah we'll try that, although at that point it might be simpler to just disable caching entirely (direct to `/dev/null`). I'll post back on how it goes.

**Comment by rjgwiz** [comment]
Setting the cache dir to `/dev/null` (e.g. via `cache-directory` in `drush.yml` or `DRUSH_PATHS_CACHE_DIRECTORY` env var) did not work for me. It failed as a `$candidate` in `DrushConfig->cache()` because `mkdir()`/`is_dir()` took it as a literal path, then drush jumped to the next candidate 

Wasn't planning on using that long term but wanted to share the info

**Comment by greg-1-anderson** [comment]
Yeah it would probably be better to have a separate config variable to disable the cache, if this is in fact a use-case we want to support.

**Comment by weitzman** [comment]
Feel free to open a new issue for disabling the cache.

**Comment by weitzman** [comment]
Follow up at #4535
